src/library/autodj/autodjprocessor.cpp

```diff
@@ -817,61 +998,196 @@ void AutoDJProcessor::calculateTransition(DeckAttributes* pFromDeck,
+        //
+        // If only the outro or intro length is marked but not both, use the one
+        // that is marked for the transition time. If neither is marked, fall
+        // back to the transition time from the spinbox.
+        if (outroLength > 0 && introLength > 0) {
+            pFromDeck->fadeBeginPos = outroStart;
+            if (outroLength > introLength) {
+                // Cut off end of outro
+                pFromDeck->fadeEndPos = outroStart + introLength;
+            } else {
+                pFromDeck->fadeEndPos = outroEnd;
+            }
+            pToDeck->startPos = introStart;
+        } else if (outroLength > 0 && introLength <= 0) {
+            pFromDeck->fadeBeginPos = outroStart;
+            pFromDeck->fadeEndPos = outroEnd;
+            pToDeck->startPos = introStart;
+        } else if (introLength > 0 && outroLength <= 0) {
+            pFromDeck->fadeBeginPos = outroEnd - introLength;
+            pFromDeck->fadeEndPos = outroEnd;
+            pToDeck->startPos = introStart;
         } else {
-            pFromDeck->posThreshold = 1.0;
-            pFromDeck->fadeDuration = 0.0;
+            useFixedFadeTime(pFromDeck, pToDeck, outroEnd, introStart);
         }
+        break;
+    case TransitionMode::FixedSkipSilence:
+        useFixedFadeTime(pFromDeck, pToDeck,
+                getLastSoundPosition(pFromDeck),
+                getFirstSoundPosition(pToDeck));
+        break;
+    case TransitionMode::FixedFullTrack:
+    default:
+        useFixedFadeTime(pFromDeck, pToDeck, fromTrackDuration, 0);
+    }
 
-        if (toTrackDuration > 0.0) {
-            // Transition time is negative here, so we need to add instead of subtract.
-            pToDeck->startPos = (toTrackIntroStart + m_transitionTime) / toTrackDuration;
-        } else {
-            pToDeck->startPos = 0.0;
+    // These are expected to be a fraction of the track length.
+    pFromDeck->fadeBeginPos /= fromTrackDuration;
+    pFromDeck->fadeEndPos /= fromTrackDuration;
+    pToDeck->startPos /= toTrackDuration;
+
+    pFromDeck->isFromDeck = true;
+    pToDeck->isFromDeck = false;
+
+    DEBUG_ASSERT(pFromDeck->fadeBeginPos <= 1);
+
+    if (sDebug) {
+        qDebug() << this << pFromDeck->fadeBeginPos << pFromDeck->fadeEndPos
+                << pToDeck->startPos;
+    }
+}
+
+void AutoDJProcessor::useFixedFadeTime(DeckAttributes* pFromDeck,
+        DeckAttributes* pToDeck, double endPoint, double startPoint) {
+    if (m_transitionTime > 0.0) {
+        // Guard against the next track being too short. This transition must finish
+        // before the next transition starts.
+        double toDeckOutroStart = getOutroStartPosition(pToDeck);
+        if (toDeckOutroStart <= startPoint) {
+            // we are already too late
+            double end = getOutroEndPosition(pToDeck);
+            if (end <= startPoint) {
+                end = pToDeck->duration();
+                VERIFY_OR_DEBUG_ASSERT(end > startPoint) {
```

Comment:
This happens if the too deck is playing and reaching the end. So it cannot be an assertion.

Reply:
https://github.com/mixxxdj/mixxx/pull/2103#issuecomment-535760843

Reply:
What do you mean with the link?

Reply:
Just what I said: I don't understand what you want to do with this function. Please make a PR for my branch to fix it.

Reply:
OK, a PR is in progress